src/panels/lovelace/cards/hui-media-control-card.ts

```diff
@@ -59,6 +59,103 @@ function getContrastRatio(
   return Math.round((contrast(rgb1, rgb2) + Number.EPSILON) * 100) / 100;
 }
 
+// For debug purposes, is being tree shaken.
+const DEBUG_COLOR = __DEV__ && false;
+
+const logColor = (
+  color: Swatch,
+  label: string = `${color.getHex()} - ${color.getPopulation()}`
+) =>
+  // tslint:disable-next-line:no-console
+  console.log(
+    `%c${label}`,
+    `color: ${color.getBodyTextColor()}; background-color: ${color.getHex()}`
+  );
+
+const customGenerator = (colors: Swatch[]) => {
+  colors.sort((colorA, colorB) => colorB.population - colorA.population);
+
+  const backgroundColor = colors[0];
+  let foregroundColor: string | undefined;
+
+  const contrastRatios = new Map<Swatch, number>();
+  const approvedContrastRatio = (color: Swatch) => {
+    if (!contrastRatios.has(color)) {
+      contrastRatios.set(
+        color,
+        getContrastRatio(backgroundColor.rgb, color.rgb)
+      );
+    }
+
+    return contrastRatios.get(color)! > CONTRAST_RATIO;
+  };
+
+  // We take each next color and find one that has better contrast.
+  for (let i = 1; i < colors.length && foregroundColor === undefined; i++) {
+    // If this color matches, score, take it.
+    if (approvedContrastRatio(colors[i])) {
+      if (DEBUG_COLOR) {
+        logColor(colors[i], "PICKED");
+      }
+      foregroundColor = colors[i].hex;
+      break;
+    }
+
+    // This color has the wrong contrast ratio, but it is the right color.
+    // Let's find similar colors that might have the right contrast ratio
+
+    const currentColor = colors[i];
+    if (DEBUG_COLOR) {
+      logColor(colors[i], "Finding similar color with better contrast");
+    }
+
+    for (let j = i + 1; j < colors.length; j++) {
+      const compareColor = colors[j];
+
+      // difference. 0 is same, 765 max difference
+      const diffScore =
+        Math.abs(currentColor.rgb[0] - compareColor.rgb[0]) +
+        Math.abs(currentColor.rgb[1] - compareColor.rgb[1]) +
+        Math.abs(currentColor.rgb[2] - compareColor.rgb[2]);
+
+      if (DEBUG_COLOR) {
+        logColor(colors[j], `${colors[j].hex} - ${diffScore}`);
+      }
+
+      if (diffScore > 150) {
+        continue;
+      }
+
+      if (approvedContrastRatio(compareColor)) {
+        if (DEBUG_COLOR) {
+          logColor(compareColor, "PICKED");
+        }
+        foregroundColor = compareColor.hex;
+        break;
+      }
+    }
+  }
+
+  if (foregroundColor === undefined) {
+    foregroundColor = colors[0].bodyTextColor;
```

Comment:
For clarity:
```suggestion
    foregroundColor = backgroundColor.bodyTextColor;
```

This always has the right contrast I assume?